File: frontend/src/api/checklist.ts
import apiClient from './client'

export type ChecklistBlock =
  | 'carryover'
  | 'entry'
  | 'payroll'
  | 'bank_visit'
  | 'statement'
  | 'cash_and_deposits'
  | 'closing'

export interface ChecklistStep {
  key: string
  block: ChecklistBlock
  /** Done outside the application — nothing here can observe it. */
  external: boolean
  signal: string | null
  route: string | null
  checked: boolean
  checked_by: string | null
  checked_at: string | null
  /** Left unchecked when the previous session was closed. */
  carried_over: boolean
}

export interface ChecklistSession {
  id: number
  period_type: string
  period: string
  status: 'open' | 'closed'
  opened_at: string
  opened_by: string | null
  closed_at: string | null
  closed_by: string | null
}

/** Observed facts, keyed by signal name. Shown next to a step, never ticking it. */
export type ChecklistSignals = Record<string, Record<string, string | number | null>>

export interface ChecklistSessionDetail {
  session: ChecklistSession
  steps: ChecklistStep[]
  signals: ChecklistSignals
}

export interface ChecklistCurrent {
  detail: ChecklistSessionDetail | null
  suggested_period: string
  checked_count: number
  total_count: number
}

export async function getCurrentChecklist(): Promise<ChecklistCurrent> {
  const response = await apiClient.get<ChecklistCurrent>('/api/checklist/current')
  return response.data
}

export async function openChecklistSession(period?: string): Promise<ChecklistSessionDetail> {
  const response = await apiClient.post<ChecklistSessionDetail>('/api/checklist/sessions', {
    period: period ?? null,
  })
  return response.data
}

export async function setChecklistStep(
  sessionId: number,
  stepKey: string,
  checked: boolean,
): Promise<ChecklistSessionDetail> {
  const response = await apiClient.put<ChecklistSessionDetail>(
    `/api/checklist/sessions/${sessionId}/steps/${stepKey}`,
    { checked },
  )
  return response.data
}

export async function closeChecklistSession(sessionId: number): Promise<ChecklistSession> {
  const response = await apiClient.post<ChecklistSession>(
    `/api/checklist/sessions/${sessionId}/close`,
  )
  return response.data
}

export async function listChecklistSessions(): Promise<ChecklistSession[]> {
  const response = await apiClient.get<ChecklistSession[]>('/api/checklist/sessions')
  return response.data
}

File: backend/schemas/checklist.py
"""Pydantic schemas for the monthly bookkeeping checklist."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from backend.models.checklist import ChecklistPeriodType, ChecklistSessionStatus

_PERIOD_PATTERN = r"^\d{4}-(0[1-9]|1[0-2])$"


class ChecklistStepRead(BaseModel):
    """One step of the checklist, with its state in the session at hand."""

    key: str
    block: str
    external: bool
    signal: str | None
    route: str | None
    checked: bool
    checked_by: str | None
    checked_at: datetime | None
    #: Left unchecked when the previous session was closed.
    carried_over: bool


class ChecklistSessionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    period_type: ChecklistPeriodType
    period: str
    status: ChecklistSessionStatus
    opened_at: datetime
    opened_by: str | None
    closed_at: datetime | None
    closed_by: str | None


class ChecklistSessionDetail(BaseModel):
    session: ChecklistSessionRead
    steps: list[ChecklistStepRead]
    #: Observed facts keyed by signal name — shown next to a step, never ticking it.
    signals: dict[str, dict[str, Any]]


class ChecklistCurrent(BaseModel):
    """What the header button needs, whether or not a session is open."""

    #: None when no session is open — the frontend then offers to start one.
    detail: ChecklistSessionDetail | None
    #: The period a new session would be about, given today's date.
    suggested_period: str
    checked_count: int
    total_count: int


class ChecklistSessionOpen(BaseModel):
    period: str | None = Field(default=None, pattern=_PERIOD_PATTERN)


class ChecklistStepUpdate(BaseModel):
    checked: bool
